src/devices/bus/spectrum/zxbus.h

```diff
@@ -81,24 +81,37 @@ DECLARE_DEVICE_TYPE(ZXBUS_SLOT, zxbus_slot_device)
 
 
 class zxbus_device : public device_t
+                   , public device_memory_interface
 {
 public:
 	zxbus_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
 
 	template <typename T> void set_iospace(T &&tag, int spacenum) { m_iospace.set_tag(std::forward<T>(tag), spacenum); }
-	template<typename T> void install_device(offs_t addrstart, offs_t addrend, T &device, void (T::*map)(class address_map &map), uint64_t unitmask = ~u64(0))
+
+	template<typename T> void install_device(offs_t addrstart, offs_t addrend, T &device, void (T::*map)(class address_map &map), u64 unitmask = ~u64(0))
 	{
 		m_iospace->install_device(addrstart, addrend, device, map, unitmask);
 	}
 
+	u8 io_r(offs_t offset);
+	void io_w(offs_t offset, u8 data);
+
 	void add_slot(zxbus_slot_device &slot);
+	void set_custom_spaces() { m_allocspaces = true; }
```

Comment:
Is adding `device_memory_interface` here the best approach, or would it be better for the system using it to create a lightweight device with an address space to use?  For example the driver can instantiate an `address_map_bank_device` just for the purpose of attaching the `zxbus_device` to its address space, and ignore its banking features entirely.  That would reduce complexity and potential for misconfiguration in this device.

Reply:
I'm thinking maybe zxbus can be this device? It's fit perfectly for this purpose. Just need some time to refactor existing drivers to use zxbus this way and then we can get rid of this "custom" spaces.

The only challenge I have now is how to prioritize customizes and make driver to contribute first to this spaces. Currently driver adds it's configuration last.

Reply:
> Is adding `device_memory_interface` here the best approach, or would it be better for the system using it to create a lightweight device with an address space to use? For example the driver can instantiate an `address_map_bank_device` just for the purpose of attaching the `zxbus_device` to its address space, and ignore its banking features entirely. That would reduce complexity and potential for misconfiguration in this device.

That is really awesome. Thank you a lot for the tip!